lib/auth/assist/assistv1/service.go

```diff
@@ -92,6 +130,67 @@ func (a *Service) CreateAssistantMessage(ctx context.Context, req *assist.Create
 }
 
 // IsAssistEnabled returns true if the assist is enabled or not on the auth level.
-func (a *Service) IsAssistEnabled(ctx context.Context, req *assist.IsAssistEnabledRequest) (*assist.IsAssistEnabledResponse, error) {
+func (a *Service) IsAssistEnabled(ctx context.Context, _ *assist.IsAssistEnabledRequest) (*assist.IsAssistEnabledResponse, error) {
+	if a.embedder == nil {
+		// If the embedder is not configured, the assist is not enabled as we cannot compute embeddings.
+		return &assist.IsAssistEnabledResponse{Enabled: false}, nil
+	}
+
+	// Check if assist can use the backend.
 	return a.backend.IsAssistEnabled(ctx)
 }
+
+func (a *Service) GetAssistantEmbeddings(ctx context.Context, msg *assist.GetAssistantEmbeddingsRequest) (*assist.GetAssistantEmbeddingsResponse, error) {
+	// TODO(jakule): The kind needs to be updated when we add more resources.
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindNode, types.VerbRead, types.VerbList)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if a.embedder == nil {
+		return nil, trace.BadParameter("assist is not configured in auth server")
+	}
+
+	// Call the openAI API to get the embeddings for the query.
+	embeddings, err := a.embedder.ComputeEmbeddings(ctx, []string{msg.ContentQuery})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(embeddings) == 0 {
+		return nil, trace.NotFound("OpenAI embeddings returned no results")
+	}
+
+	// Use default values for the id and content, as we only care about the embeddings.
+	queryEmbeddings := ai.NewEmbedding(msg.Kind, "", embeddings[0], [32]byte{})
+	documents := a.embeddings.GetRelevant(queryEmbeddings, int(msg.Limit), func(id string, embedding *ai.Embedding) bool {
+		// Run RBAC check on the embedded resource.
+		node, err := a.resourceGetter.GetNode(ctx, defaults.Namespace, embedding.GetEmbeddedID())
+		if err != nil {
+			a.log.Tracef("failed to get node %q: %v", embedding.GetName(), err)
+			return false
+		}
+		return authCtx.Checker.CheckAccess(node, services.AccessState{MFAVerified: true}) == nil
+	})
+
+	protoDocs := make([]*assist.EmbeddedDocument, 0, len(documents))
+	for _, doc := range documents {
+		node, err := a.resourceGetter.GetNode(ctx, defaults.Namespace, doc.GetEmbeddedID())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		content, err := ai.SerializeNode(node)
```

Comment:
Non-blocking: I have no idea whether yaml serialization is the right tool here.

When prototyping, I instinctively used CSV serialization when feeding nodes to the prompt in order to minimize token usage. I have no evidence CSV is better, I just followed other projects feedback that table-like structures were better understood. YAML is simpler to implement with the current appraoch though (building a CSV table requires knowing all labels, which doesn't work really well with row-by-row APIs like this one).

If we observe too high token usage and/or inaccuracies, we might want to change this, and maybe how the enpdoint works as well.

Reply:
fyi from experience, CSV is usually king for GPT-4 understanding if usable

Reply:
I changed the implementation to CSV and I see a way worst performance. Previously LLM was almost always right. Now, it's the opposite. I'll leave the YAML for now. We can try to tweak it later.